internal/controller/controllers/csr_utils.go

```diff
@@ -0,0 +1,228 @@
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	nodeUser          = "system:node"
+	nodeGroup         = "system:nodes"
+	nodeUserPrefix    = nodeUser + ":"
+	maxAgentClockSkew = 2 * time.Minute
+	maxPendingDelta   = time.Hour
+)
+
+func isNodeReady(node *corev1.Node) bool {
+	if node != nil {
+		for _, c := range node.Status.Conditions {
+			if c.Status == corev1.ConditionTrue && c.Type == corev1.NodeReady {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getAgentHostname(agent *aiv1beta1.Agent) string {
+	if agent.Spec.Hostname != "" {
+		return agent.Spec.Hostname
+	}
+	return agent.Status.Inventory.Hostname
+}
+
+func getAgentIPs(agent *aiv1beta1.Agent) (ret []string) {
+	for _, intf := range agent.Status.Inventory.Interfaces {
+		for _, addrCidr := range append(intf.IPV4Addresses, intf.IPV6Addresses...) {
+			if ip, _, err := net.ParseCIDR(addrCidr); err == nil && ip != nil {
+				ret = append(ret, ip.String())
+			}
+		}
+	}
+	return
+}
+
+func getNodeDNSNames(node *corev1.Node) (ret []string) {
+	for _, addr := range node.Status.Addresses {
+		switch addr.Type {
+		case corev1.NodeInternalDNS, corev1.NodeExternalDNS, corev1.NodeHostName:
+			ret = append(ret, addr.Address)
+		default:
+		}
+	}
+	return
+}
+
+func getX509ParsedRequest(csr *certificatesv1.CertificateSigningRequest) (*x509.CertificateRequest, error) {
+	decodedCSR, _ := pem.Decode(csr.Spec.Request)
+	if decodedCSR == nil {
+		return nil, errors.Errorf("Failed to decode request of CSR %s", csr.Name)
+	}
+	cr, err := x509.ParseCertificateRequest(decodedCSR.Bytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to parse CSR %s", csr.Name)
+	}
+	return cr, nil
+}
+
+func hasExactUsages(csr *certificatesv1.CertificateSigningRequest, usages []certificatesv1.KeyUsage) bool {
+	if len(usages) != len(csr.Spec.Usages) {
+		return false
+	}
+
+	usageMap := map[certificatesv1.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range csr.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isReqFromNodeBootstrapper(req *certificatesv1.CertificateSigningRequest) bool {
+	nodeBootstrapperUsername := "system:serviceaccount:openshift-machine-config-operator:node-bootstrapper"
+	nodeBootstrapperGroups := sets.NewString(
+		"system:serviceaccounts:openshift-machine-config-operator",
+		"system:serviceaccounts",
+		"system:authenticated",
+	)
+
+	return req.Spec.Username == nodeBootstrapperUsername && nodeBootstrapperGroups.Equal(sets.NewString(req.Spec.Groups...))
+}
+
+// Type representing a function that validates CSR for client or server.
+type nodeCsrValidator func(agent *aiv1beta1.Agent, csr *certificatesv1.CertificateSigningRequest, x509cr *x509.CertificateRequest) (bool, error)
+
+// Recognize if the CSR is a client CSR, that is sent to allow the node to join the cluster
```

Comment:
Could you please document the various checks this function does?

Reply:
OK